rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,962 @@
+  }
+
+  NamesAndTypes topics = __get_names_and_types_from_gids(
+    entities_map,
+    get_entities_gids(*node_info_ptr),
+    demangle_topic,
+    demangle_type);
+
+  return __populate_rmw_names_and_types(
+    topics,
+    allocator,
+    topic_names_and_types);
+}
+
+static
+const GraphCache::GidSeq &
+__get_writers_gids(const rmw_dds_common::msg::NodeEntitiesInfo & node_info)
+{
+  return node_info.writer_gid_seq;
+}
+
+rmw_ret_t
+GraphCache::get_writer_names_and_types_by_node(
+  const std::string & node_name,
+  const std::string & namespace_,
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type,
+  rcutils_allocator_t * allocator,
+  rmw_names_and_types_t * topic_names_and_types) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  return __get_names_and_types_by_node(
+    participants_,
+    data_writers_,
+    node_name,
+    namespace_,
+    demangle_topic,
+    demangle_type,
+    __get_writers_gids,
+    allocator,
+    topic_names_and_types);
+}
+
+static
+const GraphCache::GidSeq &
+__get_readers_gids(const rmw_dds_common::msg::NodeEntitiesInfo & node_info)
+{
+  return node_info.reader_gid_seq;
+}
+
+rmw_ret_t
+GraphCache::get_reader_names_and_types_by_node(
+  const std::string & node_name,
+  const std::string & namespace_,
+  DemangleFunctionT demangle_topic,
+  DemangleFunctionT demangle_type,
+  rcutils_allocator_t * allocator,
+  rmw_names_and_types_t * topic_names_and_types) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  return __get_names_and_types_by_node(
+    participants_,
+    data_readers_,
+    node_name,
+    namespace_,
+    demangle_topic,
+    demangle_type,
+    __get_readers_gids,
+    allocator,
+    topic_names_and_types);
+}
+
+static
+size_t
+__get_number_of_nodes(const GraphCache::ParticipantToNodesMap & participants_map)
+{
+  size_t nodes_number = 0;
+  for (const auto & elem : participants_map) {
+    nodes_number += elem.second.size();
+  }
+  return nodes_number;
+}
+
+size_t
+GraphCache::get_number_of_nodes() const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  return __get_number_of_nodes(participants_);
+}
+
+rmw_ret_t
+GraphCache::get_node_names(
+  rcutils_string_array_t * node_names,
+  rcutils_string_array_t * node_namespaces,
+  rcutils_allocator_t * allocator) const
+{
+  std::lock_guard<std::mutex> guard(mutex_);
+  if (rmw_check_zero_rmw_string_array(node_names) != RMW_RET_OK) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (rmw_check_zero_rmw_string_array(node_namespaces) != RMW_RET_OK) {
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(allocator, "get_node_names allocator is not valid",
+    return RMW_RET_INVALID_ARGUMENT);
+
+  size_t nodes_number = __get_number_of_nodes(participants_);
+  rcutils_ret_t rcutils_ret =
+    rcutils_string_array_init(node_names, nodes_number, allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    rcutils_reset_error();
```

Comment:
@ivanpauno hmm, careful with error handling here. RMW error handling [is actually `rcutils`' under the hood](https://github.com/ros2/rmw/blob/master/rmw/include/rmw/error_handling.h) so here you're just dropping the message. Same elsewhere.

Reply:
Thanks! I think that my intention was doing that in the oposite order.

Reply:
If you reset the error first then you won't have an error string to set afterwards.

Reply:
:man_facepalming: Fixed in https://github.com/ros2/rmw_dds_common/pull/4/commits/6247fc72227f2f483682673f714e381c3213a936.